cowbird/handlers/impl/filesystem.py

```diff
@@ -1,79 +1,449 @@
 import os
+import re
 import shutil
-from typing import Any
+from pathlib import Path
+from typing import Any, List, Tuple, cast
 
+from magpie.permissions import Access
+from magpie.permissions import Permission as MagpiePermission
+from magpie.services import ServiceAPI
+
+from cowbird.handlers import HandlerFactory
 from cowbird.handlers.handler import HANDLER_WORKSPACE_DIR_PARAM, Handler
+from cowbird.monitoring.fsmonitor import FSMonitor
+from cowbird.monitoring.monitoring import Monitoring
 from cowbird.permissions_synchronizer import Permission
-from cowbird.typedefs import SettingsType
-from cowbird.utils import get_logger
+from cowbird.typedefs import JSON, SettingsType
+from cowbird.utils import apply_new_path_permissions, get_logger
 
 LOGGER = get_logger(__name__)
 
 NOTEBOOKS_DIR_NAME = "notebooks"
+SECURE_DATA_PROXY_NAME = "secure-data-proxy"
+USER_WPS_OUTPUTS_USER_DIR_NAME = "wpsoutputs"
 
 
-class FileSystem(Handler):
+class FileSystem(Handler, FSMonitor):
     """
     Keep the proper directory structure in sync with the platform.
     """
     required_params = [HANDLER_WORKSPACE_DIR_PARAM]
 
-    def __init__(self, settings: SettingsType, name: str, jupyterhub_user_data_dir: str, **kwargs: Any) -> None:
+    def __init__(self,
+                 settings: SettingsType,
+                 name: str,
+                 jupyterhub_user_data_dir: str,
+                 wps_outputs_dir: str,
+                 **kwargs: Any) -> None:
         """
         Create the file system instance.
 
         :param settings: Cowbird settings for convenience
         :param name: Handler name
         :param jupyterhub_user_data_dir: Path to the JupyterHub user data directory,
                                          which will be symlinked to the working directory
+        :param wps_outputs_dir: Path to the wps outputs directory
```

Comment:
I had made it that way, because I saw the variable `wps_outputs_public_subdir` as optional. The other explicit params are required in the `config.yml` file and `kwargs` params can be optionally used in the `config.yml`.

The reason I saw this parameter as more optional, is that it is only managed by the cowbird service and I saw less necessity of it being customized for different setups. It is for a directory that is managed by the Cowbird service only, as opposed to for example `jupyterhub_user_data_dir` and `wps_outputs_dir` which are used by other services (JupyterHub, Weaver, and maybe others).

Reply:
Adding it as an actual kw parameter will make it such that you don't need to do `kwargs.get("<kw>", None)` for each of those variables that can be passed as input anyway, and it would make the class slightly better documented. 

You can still leave them optional with `wps_outputs_public_subdir: Optional[str] = None`.

Reply:
I didn't think about passing the `DEFAULT_...` values directly, but that's an even better idea! 
Given that, the type would actually be `str` directly. 
Is schema validation performed to ensure those parameters can be `str` only. 
Otherwise, a config that explicitly sets `notebooks_dir_name: null` could technically pass a `None` as parameter.

Reply:
I will update the schema validation, which didn't check for those new optional parameters, also added a restriction that the string should not be empty, to prevent weird cases.